2_process/src/training_data.py

```diff
@@ -0,0 +1,165 @@
+# Read lake-specific sequences from .npy files and save .npz files for training and test data
+
+import numpy as np
+import pandas as pd
+
+def split_indices(num_items, split_fractions, seed=0):
+    """
+    Randomly split a set into non-overlapping subsets by fractions.
+    Provide the indices for each new subset.
+    Useful for determining a Train/Test or Train/Validate/Test split
+
+    :param num_items: Number of items to split into subsets.
+    :param split_fractions: An array-like object containing a fraction for each subset. The elements of split_fractions must sum to 1.
+        E.g., for a 60-20-20 split, split_fractions=[.6, .2, .2]
+    :param seed: Random seed used to shuffle indices
+    :returns: A list of arrays of indices, one for each subset.
+
+    """
+    # Convert split_fractions to 1D numpy array
+    split_fractions = np.array(split_fractions).flatten()
+    num_splits = len(split_fractions)
+
+    # Make sure the fractions sum to 1
+    fraction_sum = split_fractions.sum()
+    # Allow for floating point precision error by using np.isclose
+    if not np.isclose(fraction_sum, 1.0):
+        raise ValueError("split_fractions must sum to 1")
+
+    # Create an array of indices from 0 to num_items - 1
+    item_indices = np.arange(num_items)
+    # Shuffle its order
+    rng = np.random.default_rng(seed)
+    rng.shuffle(item_indices)
+
+    # Determine where to split the array item_indices so that each split has
+    # the correct fraction of the total items.
+    # Running sum of split fractions
+    cumulative_fractions = split_fractions.cumsum()
+    # Multiply the running sum by the total number of items and round to an
+    # integer to get the ending index of each split
+    where_to_split = (num_items*cumulative_fractions).round().astype(int)
+    # Insert 0 at the beginning so that we can use the array for slicing
+    where_to_split = np.insert(where_to_split, 0, 0)
+    # The first split will be where_to_split[0]:where_to_split[1]
+    # The second split will be where_to_split[1]:where_to_split[2]
+    # And so on.
+
+    # Split item_indices, using where_to_split to slice
+    split_indices = [item_indices[where_to_split[i_split]:where_to_split[i_split+1]] 
+                     for i_split in range(num_splits)]
+    return split_indices
+
+
+def read_sequences(lake_sequence_files, n_depths):
+    """
+    Read a list of sequence files in, remove sequences that have any
+    NaN values among their inputs, and return a single numpy array
+    containing all sequences from all sequence files.
+
+    :param lake_sequence_files: List of lake sequence .npy files
+    :param n_depths: Number of discrete depths at which LSTM will
+    output temperatures
+    :returns: numpy array of sequences
+
+    """
+    sequences = np.concatenate([np.load(sfn) for sfn in lake_sequence_files])
+    # remove inputs with nans
```

Comment:
Maybe make clear here that you're removing inputs with nans for _features_. At first I thought you were again checking for nans within the observation data.

Reply:
Ah, I see. Yes I was thinking of the observations also as inputs to the training process. But another ML person would likely have the same interpretation as you do. I think 'input features' sounds like a good clarification.